openedx/openedx-platform
# fix: refactor MakoService to allow specifying namespace per template

## Description

Currently, the "Manage Access" menu in Studio is not working consistently (sometimes it does work though), due to a template error:

![Screenshot 2023-08-18 at 10 47 28 AM](https://github.com/openedx/edx-platform/assets/945577/9be0e8bb-2619-455f-bfc2-c6366895b76e)

This is not the first error we've had with the MakoService being unable to find mako templates, and the problem stems from different search paths for "LMS templates" vs "CMS templates", and the fact that the runtime has [this internal state](https://github.com/openedx/edx-platform/blob/39e042cfbd56accf5c1d5330da358faa9909575c/common/djangoapps/edxmako/services.py#L15-L24) that determines which paths to search for any given template. As you can see, the CMS runtime is often using the wrong paths, resulting in this error.

I believe that the best fix for this issue that will keep it from happening in the future, as it has happened at least 3 or 4 times so far, is to change the `MakoService` API to allow specifying the search path along with the template name. You want to render a template that's in `lms/templates/...`? Great, now call `render_lms_template()`. Want to render one that's in `cms/templates/...`? Great, now call `render_cms_template()`.

Of course mako templates are terrible and they should all be replaced with django templates or React UI, but that's a battle for another day.

## Supporting information

One of the past bugs: https://github.com/openedx/edx-platform/pull/32895

## Testing instructions

1. Create a course in Studio, set it's "Advanced Module List" to `["annotatable", "poll", "library_content", "lti_consumer"]` .
2. Add to the course: one of every "Advanced" XBlock listed, one each of library content, problem, text, video, and discussion.
3. Make sure the studio_view ("Edit") and author view are working (for library content, configure a library then click "View >" - that's the author view; other blocks, the author view is just what you see in studio, and usually identical to the student view). 
4. Publish the unit and view it in the LMS to make sure all the blocks are working there too.

## Deadline

ASAP - CAT-2 issue

## Other info

Private-ref: MNG-3884
